keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2346 @@
+permanently assigned to a node which can't fit the pod because of the pod's
+other resource requirements.
+
+### Sharing a single ResourceClaim
+
+Pods reference resource claims in a new PodSpec.ResourceClaims list. Each
+resource in that list can then be made available to one or more containers in
+that Pod. Depending on the capabilities defined in the
+ResourceClaim by the driver, a ResourceClaim can be used exclusively
+by one pod at a time or an unlimited number of pods. Support for additional
+constraints (maximum number of pods, maximum number of nodes) could be
+added once there are use cases for those.
+
+Users of a ResourceClaim don't need to be Pods. This KEP specifically supports
+Pods as users and describes how kube-scheduler and kubelet will deal with Pods
+that depend on a ResourceClaim, but the API and some custom resource driver
+might also be useful for controllers to manage resources without using those
+resources for Pods.
+
+### Ephemeral vs. persistent ResourceClaims lifecycle
+
+A persistent ResourceClaim has a lifecyle that is independent of any particular
+pod. It gets created and deleted by the user. This is useful for resources
+which are expensive to configure and that can be used multiple times by pods,
+either at the same time or one after the other. Such persistent ResourceClaims
+get referenced in the pod spec by name. When a PodTemplateSpec in an app
+controller spec references a ResourceClaim by name, all pods created by that
+controller also use that name and thus share the resources allocated for that
+ResourceClaim.
+
+But often, each Pod is meant to have exclusive access to its own ResourceClaim
+instance instead. To support such ephemeral resources without having to modify
+all controllers that create Pods, an entry in the new PodSpec.ResourceClaims
+list can also be a ResourceClaimTemplate. When a Pod gets created, such a
+template will be used to create a normal ResourceClaim with the Pod as owner
+with an
+[OwnerReference](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#OwnerReference)),
+and then the normal allocation of the resource takes place. Once the pod got
+deleted, the Kubernetes garbage collector will also delete the
+ResourceClaim.
+
+The difference between persistent and ephemeral resources for kube-scheduler
+and kubelet is that the name of the ResourceClaim needs to be determined
+differently: the name of an ephemeral ResourceClaim is `<pod>-<resource name in
+the pod spec>`. Ownership must be checked to detect accidental conflicts with
+persistent ResourceClaims or previous incarnations of the same ephemeral
+resource. This is the same approach that was chosen for [generic ephemeral
+volumes](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1698-generic-ephemeral-volumes/README.md).
+For a resource driver there is no difference.
+
+### Coordinating resource allocation through the scheduler
```

Comment:
+1 on marking this section as `UNRESOLVED` and the KEP as `provisional` so we can iterate the KEP on a new PR dedicated to scheduling and autoscaling.

I'm on the fence on the idea of prototyping. It's not really valuable for me if I'm not looking at it. In other words, if @pohly has changes in a branch, it's not easy for me to provide feedback. Unless we create a draft PR based on it. But then I would like multiple PRs, one dedicated to scheduling.

Unfortunately, github doesn't provide the ability to create chains of PRs, so maybe we could do something like this:
1. have a branch in pohly/kubernetes (could it be in k/k?)
2. have reviewers fully review the parts that are already agreed on.
3. have a PR for scheduling that we can review and iterate on.
4. Iterate KEP
5. Create PRs for k/k:master once we have agreed.

Reply:
So if we "only" merge it as provisional, no code will go into 1.25, right?

> This will hopefully let us merge the parts we like and start a new PR on the rest 

I was confused about this until I realized that you meant merging "parts of the KEP", not "parts of the code", right?

>  Also, a KEP is not a full API review, and APIs regularly change as the implementation finds gaps

So, do another prototype to verify the revised API?

Reply:
> So if we "only" merge it as provisional, no code will go into 1.25, right?

while `provisional`, no. But, theoretically, you could still make it to `implementable` in the next days.

Reply:
After thinking about this some more, I'm getting to these conclusions:

- The KEP as it stands right now could be implemented, including the part for autoscaling. I can prepare a prototype to demonstrate that. But that doesn't help when the API isn't as intended.

- I can't say the same about a solution where some state of a claim is in one object and some other state is in some other. We first need to discuss how to handle the new corner cases and race conditions that arise from that before it makes sense to prototype it. This may lead to a complete redesign of the API to prevent problematic cases. For example, "two pods share the same claim with delayed allocation" could be avoided by not creating separate ResourceClaim objects for inline resources and instead dealing with them as a separate case in all code. Or I am completely wrong and there are some simple solutions that I just don't see yet. :shrug: 

Either way, we will need a place to discuss such an alternative approach because we will either have to make it work or convince ourselves that it cannot work or would be worse than the current proposal. I'm not sure how to do this in the KEP process. If we merge this PR as "provisional", how do we continue with all the open discussion threads that we have in this PR here? Move each of them into issues that need to be resolved before allowing the provisional KEP to move to "implementable"? Where would those issues live? In this repo we would spam the entire Kubernetes community. Perhaps in https://github.com/pohly/enhancements/issues?

Or shall we go back to a Google Doc and discuss there? That still doesn't give us an issue tracker, but at least comment handling would be saner than in a GitHub PR and we could do collaborative editing of that future KEP revision.

Reply:
One downside of merging a large chunk of text that is UNRESOLVED is that we cannot comment on that existing text again in a new PR unless that PR touches the text. I'm not saying we shouldn't do it, I just wanted to point it out.

Reply:
I managed to convince myself that moving `SelectedNode` into `PodScheduling` is doable. But I hit a snag with moving `Deallocate`. Instead of putting more text into this PR with a proposal that has a known issue, let's try whether discussing that change in a separate PR works. Please see https://github.com/pohly/enhancements/pull/13. The two new races get discussed at the top, with multiple alternative solutions for the second one of which all have drawbacks.